fms/models/roberta.py

```diff
@@ -0,0 +1,331 @@
+        position_out = self.position_embedding(position_ids)
+
+        # zero out the associated position embeddings
+        if self.config.pad_id is not None:
+            position_out = position_out.mul(~is_pad.unsqueeze(-1))
+
+        # perform absolute position embedding
+        x = x_emb + position_out
+
+        # layer norm
+        x = self.enc_norm(x)
+
+        # add dropout
+        if self.config.p_dropout:
+            x = self.dropout(x)
+
+        # layers
+        for layer in self.layers:
+            x = layer(x, mask=mask)
+
+        return x
+
+
+class RoBERTa(nn.Module):
+    def __init__(self, config: Optional[RoBERTaConfig] = None, **kwargs):
+
+        super(RoBERTa, self).__init__()
+        if config is not None:
+            self.config = config
+        else:
+            self.config = RoBERTaConfig()
+        self.config = self.config.updated(**kwargs)
+
+        self.base_model = RoBERTaHeadless(self.config)
+
+        self.class_head = ClassHead(
+            self.config.emb_dim,
+            str_to_activation(self.config.activation_fn),
+            self.config.norm_eps,
+        )
+
+        self.head = nn.Linear(self.config.emb_dim, self.config.src_vocab_size)
+
+        # this model ties weights, so we tie here
+        self.head.weight = self.base_model.embedding.weight
+
+    def forward(
+        self,
+        x: torch.LongTensor,
+        mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+    ):
+        # run through the encoder layers
+        x = self.base_model(x, mask=mask, position_ids=position_ids)
+
+        # run through the class head (using the first in each sequence in the batch as the cls_token)
+        x = self.class_head(x)
+
+        # project to vocab space
+        x = self.head(x)
+        return x
+
+    @classmethod
+    def from_config(cls, config: RoBERTaConfig) -> "RoBERTa":
+        return cls(config)
+
+    def get_config(self) -> RoBERTaConfig:
+        return self.config
+
+    def reset_params(self):
+        # Modules are self-initializing, we're just going to down-scale the final prediction head to be
+        # mixed-fan (inputs and gradients scale to the same inverse factors) if it isn't tied
+        self.head.weight.data.normal_(
+            0,
+            1 / math.sqrt(math.sqrt(self.config.emb_dim * self.config.src_vocab_size)),
+        )
+
+
+# a micro llama model to use with a char-level tokenizer
+_micro_char_config = RoBERTaConfig(
+    emb_dim=192, nheads=4, nlayers=5, max_pos=1024, src_vocab_size=256
+)
+
+_base_config = RoBERTaConfig()
+
+_architecture_name = "roberta"
+
+
+def _roberta_factory_factory(config):
+    def factory(**kwargs):
+        return RoBERTa(config, **kwargs)
+
+    return factory
+
+
+models.register_model(
+    _architecture_name, "micro", _roberta_factory_factory(_micro_char_config)
+)
+models.register_model(
+    _architecture_name, "base", _roberta_factory_factory(_base_config)
```

Comment:
is this 'base' config associated with anything in particular, relative to other potential variants?

Given that RoBERTa references a training regiment applied to the BERT architecture, I wonder if the architecture should be called 'bert' (and `class BERT`) with the 'variant' called 'roberta-base'. Then again maybe there isn't really a 'config' difference between bert and roberta either, just a difference in weights. The naming for this set of models can be a bit confusing.

Reply:
Yeah, RoBERTa and BERT are the same architecturally. I'd stick with RoBERTa currently as of the two, it's the more commonly accepted representative for the family of language encoders

Reply:
Yeah, RoBERTa and BERT are the same architecturally. I'd stick with RoBERTa currently as of the two, it's the more commonly accepted representative for the family of language encoders